x/sync/manager.go

```diff
@@ -354,46 +384,267 @@ func (m *Manager) getAndApplyRangeProof(ctx context.Context, work *workItem) {
+		var changeProof merkledb.ChangeProof
+		if err := changeProof.UnmarshalProto(changeProofResp.ChangeProof); err != nil {
+			return err
+		}
+
+		// Ensure the response does not contain more than the requested number of leaves
+		// and the start and end roots match the requested roots.
+		if len(changeProof.KeyChanges) > int(request.KeyLimit) {
+			return fmt.Errorf(
+				"%w: (%d) > %d)",
+				errTooManyKeys, len(changeProof.KeyChanges), request.KeyLimit,
+			)
+		}
+
+		endRoot, err := ids.ToID(request.EndRootHash)
+		if err != nil {
+			return err
+		}
+
+		if err := m.config.DB.VerifyChangeProof(
+			ctx,
+			&changeProof,
+			startKey,
+			endKey,
+			endRoot,
+		); err != nil {
+			return fmt.Errorf("%w due to %w", errInvalidChangeProof, err)
+		}
+
+		changeOrRangeProof = &merkledb.ChangeOrRangeProof{
+			ChangeProof: &changeProof,
+		}
+	case *pb.SyncGetChangeProofResponse_RangeProof:
+
+		var rangeProof merkledb.RangeProof
+		if err := rangeProof.UnmarshalProto(changeProofResp.RangeProof); err != nil {
+			return err
+		}
+
+		// The server did not have enough history to send us a change proof
+		// so they sent a range proof instead.
+		if err := verifyRangeProof(
+			ctx,
+			&rangeProof,
+			int(request.KeyLimit),
+			startKey,
+			endKey,
+			request.EndRootHash,
+			m.tokenSize,
+			m.config.Hasher,
+		); err != nil {
+			return err
+		}
+
+		changeOrRangeProof = &merkledb.ChangeOrRangeProof{
+			RangeProof: &rangeProof,
+		}
+	default:
+		return fmt.Errorf(
+			"%w: %T",
+			errUnexpectedChangeProofResponse, changeProofResp,
+		)
 	}
 
-	if len(proof.KeyValues) > 0 {
-		largestHandledKey = maybe.Some(proof.KeyValues[len(proof.KeyValues)-1].Key)
+	if changeOrRangeProof.ChangeProof != nil {
+		// The server had sufficient history to respond with a change proof.
+		changeProof := changeOrRangeProof.ChangeProof
+		largestHandledKey := work.end
+		// if the proof wasn't empty, apply changes to the sync DB
+		if len(changeProof.KeyChanges) > 0 {
+			if err := m.config.DB.CommitChangeProof(ctx, changeProof); err != nil {
+				m.setError(err)
+				return nil
+			}
+			largestHandledKey = maybe.Some(changeProof.KeyChanges[len(changeProof.KeyChanges)-1].Key)
+		}
+
+		m.completeWorkItem(ctx, work, largestHandledKey, targetRootID, changeProof.EndProof)
+		return nil
+	}
+
+	// The server responded with a range proof.
+	rangeProof := changeOrRangeProof.RangeProof
+	largestHandledKey := work.end
+	if len(rangeProof.KeyValues) > 0 {
+		// Add all the key-value pairs we got to the database.
+		if err := m.config.DB.CommitRangeProof(ctx, work.start, work.end, rangeProof); err != nil {
+			m.setError(err)
+			return nil
+		}
+		largestHandledKey = maybe.Some(rangeProof.KeyValues[len(rangeProof.KeyValues)-1].Key)
 	}
 
-	m.completeWorkItem(ctx, work, largestHandledKey, targetRootID, proof.EndProof)
+	m.completeWorkItem(ctx, work, largestHandledKey, targetRootID, rangeProof.EndProof)
+	return nil
```

Comment:
What is the purpose of:
```golang
var changeOrRangeProof *merkledb.ChangeOrRangeProof
```
I feel like the entire `merkledb.ChangeOrRangeProof` should be removed and the usage of that variable should be replaced with moving the logic into the switch statement.

Reply:
It's not clear to me... this was previously existing code. As far as I can tell it's not needed though. We can clean it up as part of this PR.

Reply:
iirc this exists because the response to a change proof request can be either a range proof or a change proof, so we need to be able to parse either response

Reply:
> iirc this exists because the response to a change proof request can be either a range proof or a change proof, so we need to be able to parse either response

Both responses are already handled. We end up wrapping the different responses into another custom type for (afaict) no reason